src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ValidationHelpers.cs

```diff
@@ -57,50 +58,72 @@ public static string GetModelErrorMessageOrDefault(ModelError modelError)
+                // Special case the root node. It is only 'Visible` if it has any errors associated with it.
+                if (modelStateDictionary.Root.Errors.Count == 0)
+                {
+                    entries.RemoveAt(entries.Count - 1);
+                }
+
+                if (entries.Count < modelStateDictionary.Count)
+                {
+                    // Account for entries in the ModelStateDictionary that do not have corresponding ModelMetadata values.
+                    foreach (var entry in modelStateDictionary)
+                    {
+                        if (!entries.Contains(entry.Value))
+                        {
+                            entries.Add(entry.Value);
+                        }
+                    }
+                }
 
-                return viewData.ModelState
-                    .OrderBy(data => orderer.GetOrder(data.Key))
-                    .Select(ms => ms.Value);
+                return entries;
             }
+
+            return EmptyModelStateEntries;
         }
 
-        // Helper for sorting modelStates to respect the ordering in the metadata.
-        // ModelState doesn't refer to ModelMetadata, but we can correlate via the property name.
-        private class ErrorsOrderer
+        private static void Visit(
+            ModelStateDictionary dictionary,
+            ModelStateEntry modelStateEntry,
+            ModelMetadata metadata,
+            List<ModelStateEntry> orderedModelStateEntries)
         {
-            private readonly Dictionary<string, int> _ordering =
-                new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
-
-            public ErrorsOrderer(ModelMetadata metadata)
+            if (metadata.ElementMetadata != null)
             {
-                if (metadata == null)
+                foreach (var indexEntry in modelStateEntry.Children)
                 {
-                    throw new ArgumentNullException(nameof(metadata));
+                    Visit(dictionary, indexEntry, metadata.ElementMetadata, orderedModelStateEntries);
                 }
-
-                foreach (var data in metadata.Properties)
+            }
+            else if (metadata.Properties.Count == 0)
+            {
+                var descendantEntries = dictionary.FindKeysWithPrefix(metadata.PropertyName);
+                foreach (var childEntry in descendantEntries)
                 {
-                    _ordering[data.PropertyName] = data.Order;
+                    orderedModelStateEntries.Add(childEntry.Value);
                 }
             }
 
-            public int GetOrder(string key)
+            for (var i = 0; i < metadata.Properties.Count; i++)
             {
-                int value;
-                if (_ordering.TryGetValue(key, out value))
+                var propertyMetadata = metadata.Properties[i];
+                var propertyModelStateEntry = modelStateEntry.GetModelStateForProperty(propertyMetadata.PropertyName);
```

Comment:
Like the `Children` case, `GetModelStateForProperty()` will return entries for `!Visible` nodes. This works in this case. But is it really the behaviour we want to expose? Odd for `ModelStateDictionary` to have some methods / properties that expose only `Visible` nodes and some expose everything.

At least, the methods that will return `ModelStateEntry`s that were never explicitly added should have comments making this clear.